### DISCUSSION TITLE
Useless network interface created on bridge mode?

### DISCUSSION THREAD

**Discussion by FelipeGXavier** [moby/moby#47113] [Q&A]
I am currently conducting tests to gain a deeper understanding of how Docker's network bridge functions at a fundamental level. In my testing scenario, I follow these steps (using Linux):  

```docker network create my_bridge_network```

Upon inspection of my system, I observe the creation of a bridge:

```
brctl show
bridge name	  interfaces	
br-f29d3e08d992   no
```

Subsequently, by running the ```ifconfig``` command, I identify the instantiation of a networking interface named **br-f29d3e08d992** with the IP address **172.28.0.1**.

Following this setup, I create a Docker container using the previously established network bridge:

```docker run --network=my_bridge_network -d redis```

The creation of the container results in the appearance of a network interface named **veth0a4fd83**.

When attempting to ping the Docker container's address (172.28.0.2) running on the bridge, I observe that the communication traverses through the **veth0a4fd83** interface as well.

In light of this, I am curious about the purpose of Docker in creating the **br-f29d3e08d992** network interface alongside the bridge when the bridge is initially established.

**Comment by polarathene** [comment]
> I am currently conducting tests to gain a deeper understanding of how Docker's network bridge functions at a fundamental level.

If you're interested in some of the other related networking behaviour, [this resource](https://github.com/docker/docs/issues/17312#issuecomment-1736446266) may be of interest to you (_and perhaps [this one too](https://github.com/moby/moby/issues/38784#issuecomment-1565219612)_), it provides a deep dive on details. Notably a useful table for some differences with `userland-proxy` setting.

That was all in an effort to [help move the `userland-proxy: false` by default forward](https://github.com/moby/moby/issues/14856#issuecomment-1541139817) by going over known issues and verifying if they're reproducible / resolvable.

`userland-proxy: true` may be relevant to you with your goal as it has a security gotcha to be mindful of.
- If a host can be reached over IPv6, a published port from an IPv4 only container network (_when not restricting the interface to bind to_) will also allow IPv6 connections on that port to be routed to the container.
- This goes through the docker bridge networks gateway which due to iptables rules will rewrite the client/remote source IP to that gateway, leading to [problems like documented here](https://docker-mailserver.github.io/docker-mailserver/v13.3/config/advanced/ipv6/#what-can-go-wrong).

---

> The creation of the container results in the appearance of a network interface named **veth0a4fd83**.
> 
> When attempting to ping the Docker container's address (172.28.0.2) running on the bridge, I observe that the communication traverses through the **veth0a4fd83** interface as well.
> 
> In light of this, I am curious about the purpose of Docker in creating the **br-f29d3e08d992** network interface alongside the bridge when the bridge is initially established.

I think you may find [this comment](https://github.com/moby/moby/issues/44973#issuecomment-1543733757) provides information on this configuration for you? I looked into it a while back.

Without digging through all that, I think I vaguely recall it having something to do with each container is added into a network namespace and the veth is paired to the docker network bridge, which bridges to the actual host interface.

My linked comment mentions that early on with how a bridge has a limited number of "ports" (_not network ports you may be familiar with_), which as the issue identifies is the reason each bridge can only support a little over 1000 containers.
